From fa1d67827527297ebe304485023b10a1dc122dd2 Mon Sep 17 00:00:00 2001
From: Masahito S <firelzrd@gmail.com>
Date: Tue, 26 Dec 2023 01:28:37 +0900
Subject: [PATCH] lru_gen workingset protection fix

---
 mm/vmscan.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 3d98bbf354..4afad33af3 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -5652,6 +5652,8 @@ static void lru_gen_shrink_node(struct pglist_data *pgdat, struct scan_control *
 	if (!sc->may_writepage || !sc->may_unmap)
 		goto done;
 
+	lru_gen_age_node(pgdat, sc);
+
 	lru_add_drain();
 
 	blk_start_plug(&plug);
@@ -7356,7 +7358,6 @@ static void kswapd_age_node(struct pglist_data *pgdat, struct scan_control *sc)
 	struct lruvec *lruvec;
 
 	if (lru_gen_enabled()) {
-		lru_gen_age_node(pgdat, sc);
 		return;
 	}
 
-- 
2.25.1

